drivers/iio/dac/ad5413.c

```diff
@@ -0,0 +1,728 @@
+	 * Given the desired slew time, the following algorithm determines the
+	 * best match for the step size and the update clock frequency.
+	 */
+	for (i = 0; i < ARRAY_SIZE(ad5413_sr_clk); i++) {
+		/*
+		 * Go through each valid update clock freq and determine a raw
+		 * value for the step size by using the formula:
+		 * Step Size = Full Scale Out / (Update Clk Freq * Slew Time)
+		 */
+		sr_step = AD5413_FULL_SCALE_MICRO;
+		do_div(sr_step, ad5413_sr_clk[i]);
+		do_div(sr_step, st->slew_time);
+		/*
+		 * After a raw value for step size was determined, find the
+		 * closest valid match
+		 */
+		res = ad5413_find_best_match(ad5413_sr_step,
+						ARRAY_SIZE(ad5413_sr_step),
+						sr_step);
+		/* Calculate the slew time */
+		calc_slew_time = AD5413_FULL_SCALE_MICRO;
+		do_div(calc_slew_time, ad5413_sr_step[res]);
+		do_div(calc_slew_time, ad5413_sr_clk[i]);
+		/*
+		 * Determine with how many microseconds the calculated slew time
+		 * is different from the desired slew time and store the diff
+		 * for the next iteration
+		 */
+		diff_new = abs(st->slew_time - calc_slew_time);
+		if (diff_new < diff_old) {
+			diff_old = diff_new;
+			sr_clk_idx = i;
+			sr_step_idx = res;
+		}
+	}
+
+	return ad5413_slew_rate_set(st, sr_clk_idx, sr_step_idx);
+}
+
+static int ad5413_set_out_range(struct ad5413_state *st, int range)
+{
+	int ret;
+
+	ret = ad5413_spi_write_mask(st, AD5413_REG_DAC_CONFIG,
+					AD5413_REG_DAC_CONFIG_RANGE_MSK,
+					FIELD_PREP(AD5413_REG_DAC_CONFIG_RANGE_MSK, (range)));
+	if (ret < 0)
+		return ret;
+
+	/* Wait to allow time for the internal calibrations to complete */
+	return ad5413_wait_for_task_complete(st, AD5413_REG_DIGITAL_DIAG_RESULTS,
+							AD5413_REG_DIGITAL_DIAG_CAL_MEM_UNREFRESHED_MSK);
+}
+
+static int ad5413_internal_buffers_en(struct ad5413_state *st, bool enable)
+{
+	int ret;
+
+	ret = ad5413_spi_write_mask(st, AD5413_REG_DAC_CONFIG,
+					AD5413_REG_DAC_CONFIG_INT_EN_MSK,
+					FIELD_PREP(AD5413_REG_DAC_CONFIG_INT_EN_MSK, (enable)));
+	if (ret < 0)
+		return ret;
+
+	/* Wait to allow time for the internal calibrations to complete */
+	return ad5413_wait_for_task_complete(st, AD5413_REG_DIGITAL_DIAG_RESULTS,
+						AD5413_REG_DIGITAL_DIAG_CAL_MEM_UNREFRESHED_MSK);
+}
+
+static int ad5413_reset(struct ad5413_state *st)
+{
+	if (st->gpio_reset) {
+		gpiod_set_value(st->gpio_reset, 1);
+		fsleep(100);
+		gpiod_set_value(st->gpio_reset, 0);
+		fsleep(100);
+
+		return 0;
+	}
+
+	/* Perform a software reset */
+	return ad5413_soft_reset(st);
+}
+
+static int ad5413_reg_access(struct iio_dev *indio_dev,
+					unsigned int reg,
+					unsigned int writeval,
+					unsigned int *readval)
+{
+	struct ad5413_state *st = iio_priv(indio_dev);
+	int ret;
+
+	guard(mutex)(&st->lock);
+	if (readval) {
+		ret = ad5413_spi_reg_read(st, reg);
+		if (ret < 0)
+			return ret;
+
+		*readval = ret;
+		ret = 0;
+	} else {
+		ret = ad5413_spi_reg_write(st, reg, writeval);
+	}
+
+	return ret;
```

Comment:
```suggestion
		return 0;
	}
	
	return ad5413_spi_reg_write(st, reg, writeval);
```

can be simplified (direct returns are preferred upstream when possible)